designsafe/static/scripts/ng-designsafe/services/file-operation-service.js

```diff
@@ -0,0 +1,568 @@
+        });
+        const folders = files.filter(({ format }) => format === 'folder');
+        // If selection contains folders, we need to perform a listing in each to find the images
+        if (folders.length) {
+            const folderListing = (folder) =>
+                this.$http.get(
+                    this.removeDuplicateSlashes(
+                        `/api/datafiles/${api}/${scheme}/listing/${system}/${encodeURIComponent(folder.path)}/`
+                    )
+                );
+            //forkJoin emits when all folder listings have completed
+            return forkJoin(folders.map(folderListing)).pipe(
+                //tap callback pushes each image in each listing response onto the images array
+                tap((responses) => {
+                    responses.forEach((res) => {
+                        const output = res.data.listing.filter(({ path }) => {
+                            const ext = path
+                                .split('.')
+                                .pop()
+                                .toLowerCase();
+                            return ['jpg', 'jpeg', 'png', 'tiff', 'gif'].includes(ext);
+                        });
+                        images.push(...output);
+                    });
+                }),
+                //the final output of the observable is the images array
+                map((_) => images)
+            );
+        } else {
+            // If there are no folders in the selection, just emit the selected images
+            return of(images);
+        }
+    }
+
+    openImagePreviewModal({ api, scheme, system, files }) {
+        this.getPreviewableImages({ api, scheme, system, files }).subscribe((images) => {
+            var modal = this.$uibModal.open({
+                component: 'ddimagepreview',
+                resolve: {
+                    images: () => images,
+                    api: () => api,
+                    scheme: () => scheme,
+                },
+                size: 'lg',
+            });
+        });
+    }
+
+    /***************************************************************************
+                                MKDIR MODAL
+    ***************************************************************************/
+
+    openMkdirModal({ api, scheme, system, path }) {
+        var modal = this.$uibModal.open({
+            component: 'mkdirModal',
+            resolve: {
+                path: () => path,
+                api: () => api,
+                scheme: () => scheme,
+                system: () => system,
+            },
+            size: 'lg',
+        });
+    }
+
+    handleMkdir({ api, scheme, system, path, folderName, successCallback }) {
+        const mkdirUrl = `/api/datafiles/${api}/${scheme}/mkdir/${system}/${encodeURIComponent(path)}/`;
+        const mkdirRequest = this.$http.put(mkdirUrl, { dir_name: folderName }).then(successCallback);
```

Comment:
Should we be handling errors on requests?

Reply:
Because the UX flow now is to close the modals on submit and use toasts to display updates to the user, I don't know if there's actually anything we need to do on the frontend if an operation fails (as opposed to a listing).